### PR TITLE
MVC Component: Main Gallery

### LINKED ISSUE #21: MVC component: Main Gallery
image

### PR DESCRIPTION
**Pull request by MozdolevskyiMaksym**
MainGallery1
MainGallery2
MainGallery3

### CODE REVIEW

**Review by iklopov — APPROVED** [review]

**Review by MarinaPasternak — APPROVED** [review]

**Comment by Anastasiia0999** [comment]
Do we use export default? 
```
export default class ViewCarsGallery 
export default class ControllerCarsGallery
export default class ModelCarsGallery
```

**Comment by Anastasiia0999** [comment]
It`s better to split into two functions: renderGallery and renderCar
image

**Comment by Anastasiia0999** [comment]
`this.view = new ViewCarsGallery(this.handleFilterButton, this.handleShowAll, this.handleCarDetails, this.handleLoadMore);`
There are 4 callbacks here. Try to use one func instead.
getMethods = () => ({
    handleFilterButton: this.handleFilterButton,
    ...
  });

File: app/components/cars-gallery/controller-cars-gallery.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    this.model.loadCars(this.handleLoadCars);
+  }
+
+  handleLoadCars = (cars) => {                          
+    this.notify(this.events.LOAD_DATA, cars); 
+    this.view.renderStructure();  
+    this.handleLoadMore();                     
+  };
+
+  handleLoadMore = _ => {
```
Comment (PavelKundenko) [inline app/components/cars-gallery/controller-cars-gallery.js]: Just a small comment: I think it's better use "( )" than pseudo-argument "_"

**Review by PavelKundenko — APPROVED** [review]

**Comment by Anastasiia0999** [comment]
You render info for one car only => renderCar not  renderCars
`renderCars = ({photoUrl, brand, model, price, id}) => {`

File: app/components/cars-gallery/controller-cars-gallery.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    this.subscribe(this.events.SEARCH_MODEL, this.handleFilteredData);
+    this.subscribe(this.events.FILTERED_DATA, this.handleFilteredData);
+    this.subscribe(this.events.LOGO_SHOWCARS, this.handleLoadCars); 
+    
+    this.model.loadCars(this.handleLoadCars);
+  }
+
+  handleLoadCars = (cars) => {                          
+    this.notify(this.events.LOAD_DATA, cars);
+    this.model.filteredCars.length = 0;
```
Comment (so2niko) [inline app/components/cars-gallery/controller-cars-gallery.js]: This code must be named and moved to some method in model, like `clearFilteredData `or else

File: app/components/cars-gallery/controller-cars-gallery.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+  };
+
+  handleFilterButton = () => {
+    this.notify(this.events.SHOW_FILTER);        
+  };
+
+  handleShowAll = () => {                 
+    this.model.page = 1;
+    this.model.filteredCars.length = 0;
```
Comment (so2niko) [inline app/components/cars-gallery/controller-cars-gallery.js]: This code must be named and moved to some method in model, like clearFilteredData or else

File: app/components/cars-gallery/model-cars-gallery.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+export class ModelCarsGallery {
+  cars = [];
+  filteredCars = [];
+  page = 1;
+  size = 9;
+
+  loadCars(handleLoadCars) {
+    fetch('../../../base/database.json')
```
Comment (so2niko) [inline app/components/cars-gallery/model-cars-gallery.js]: move link to class property

**Review by so2niko — CHANGES_REQUESTED** [review]

**Review by Anastasiia0999 — APPROVED** [review]

File: app/components/cars-gallery/controller-cars-gallery.js (hunk 2 above)
Comment (MozdolevskyiMaksym) [reply app/components/cars-gallery/controller-cars-gallery.js]: I`v already changed it, you can check

File: app/components/cars-gallery/controller-cars-gallery.js (hunk 3 above)
Comment (MozdolevskyiMaksym) [reply app/components/cars-gallery/controller-cars-gallery.js]: It`s also done

File: app/components/cars-gallery/model-cars-gallery.js (hunk 4 above)
Comment (MozdolevskyiMaksym) [reply app/components/cars-gallery/model-cars-gallery.js]: You can also check it)

File: app/components/cars-gallery/controller-cars-gallery.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  handleShowAll = () => {                 
+    this.clearFilteredData();                       
+    const cars = this.model.getCarsPagin();
+    this.view.render(cars);
+  };
+  
+  clearFilteredData = () => {
+    this.model.page = 1;
+    this.model.filteredCars.length = 0;
```
Comment (so2niko) [inline app/components/cars-gallery/controller-cars-gallery.js]: same code...
image
Move to model, not controller

File: app/components/cars-gallery/view-cars-gallery.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    <div class="cars-gallery d-flex justify-content-around flex-wrap text-center">
+        
+    </div>
+    <div class="container-loadmore mx-auto mt-5">
+      <button type="button" class="container-loadmore__btn load-more w-100 border-0">Load more...</button>
+    </div>
+    
+  `);
+    this.addEvents();
+  }
```
Comment (so2niko) [inline app/components/cars-gallery/view-cars-gallery.js:33]: Why in some methods u use `;` at the end of line, but do not use here?

**Review by so2niko — CHANGES_REQUESTED** [review]

File: app/components/cars-gallery/view-cars-gallery.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    <div class="cars-gallery d-flex justify-content-around flex-wrap text-center">
+        
+    </div>
+    <div class="container-loadmore mx-auto mt-5">
+      <button type="button" class="container-loadmore__btn load-more w-100 border-0">Load more...</button>
+    </div>
+    
+  `);
+    this.addEvents();
+  };
```
Comment (so2niko) [inline app/components/cars-gallery/view-cars-gallery.js]: why `; `is here?

File: app/components/cars-gallery/view-cars-gallery.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  
+  render = (data) => {
+    this.carsGallery.innerHTML = data.map(this.renderCar).join('');
+  };
```
Comment (so2niko) [inline app/components/cars-gallery/view-cars-gallery.js:37]: why `; `is here?

File: app/components/cars-gallery/view-cars-gallery.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    return `
+      <div class="cars-gallery__car">
+        <img class="cars-gallery__car-img" src="${photoUrl[0]}"></img>
+        <h2 class="cars-gallery__car-brand pt-3 h2 font-weight-bold">${brand}</h2>
+        <span class="cars-gallery__car-model">${model}</span>
+        <p class="cars-gallery__car-price">from <b>₴${price}</b></p>
+        <button class="cars-gallery__car-details-btn border-0 text-primary" type="button" data-id="${id}">More information</button>
+      </div>
+    `
+  };
```
Comment (so2niko) [inline app/components/cars-gallery/view-cars-gallery.js:49]: why `; `is here?

File: app/components/cars-gallery/view-cars-gallery.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    const buttonAll = document.querySelector('.container-filter__showall-btn');
+    buttonAll.addEventListener('click', this.handleShowAll);
+
+    const buttonLoadMore = document.querySelector('.container-loadmore');
+    buttonLoadMore.addEventListener('click', this.handleLoadCars);
+
+    this.carsGallery = document.querySelector('.cars-gallery');
+    this.carsGallery.addEventListener('click', this.handleCarDetails);
+  };
```
Comment (so2niko) [inline app/components/cars-gallery/view-cars-gallery.js:63]: why `; `is here?

**Review by so2niko — CHANGES_REQUESTED** [review]

**Review by so2niko — APPROVED** [review]

**Pull request merged**